File: Algorithm/bin/slidingwindow/C2_H_StringAnagrams/C2_H_StringAnagrams.py
def findAnagrams(s, patten):
  start, match = 0, 0
  p = {}
  result = []

  for c in patten:
    if c not in p:
      p[c] = 0
    p[c] += 1
  
  for end in range(0,len(s)):
    rc = s[end]
    if rc in p:
      p[rc] -= 1
      if p[rc] == 0:
        match +=1
    
    if match == len(p):
      result.append(start)
    
    if end >= len(patten) - 1:
      lc = s[start]
      start += 1
      if lc in p:
        if p[lc] == 0:
          match -= 1
        p[lc] += 1 
  
  return result

def main():
  print(findAnagrams("ppqp", "pq"))

main()

File: Algorithm/src/slidingwindow/C2_H_StringAnagrams/C2_H_StringAnagrams.java
package slidingwindow.C2_H_StringAnagrams;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class C2_H_StringAnagrams {
  public static List<Integer> findAnagrams(String s, String pattern) {
    int start = 0, match = 0;
    Map<Character, Integer> map = new HashMap<>();

    for (char c : pattern.toCharArray()) {
      map.put(c, map.getOrDefault(c, 0) + 1);
    }
    
    List<Integer> result = new ArrayList<>();
    for(int end = 0; end < s.length(); end++){
      char rc = s.charAt(end);
      if(map.containsKey(rc)){
        map.put(rc, map.get(rc) - 1);
        if(map.get(rc) == 0){
          match++;
        }
      }

      if(match == map.size()){
        result.add(start);
      }

      if(end >= pattern.length() - 1){
        char lc = s.charAt(start);
        if(map.containsKey(lc)){
          if(map.get(lc) == 0){
            match--;
          }
          map.put(lc, map.get(lc) + 1);
        }
        start++;
      }
    }
    return result;
  }

  public static void main(String[] args) {
    System.out.println(findAnagrams("ppqp", "pq"));
    System.out.println(findAnagrams("abbcabc", "abc"));
  }
}